sql/src/main/java/io/crate/analyze/expressions/ExpressionAnalyzer.java

```diff
@@ -245,14 +255,47 @@ protected Symbol convertFunctionCall(FunctionCall node, ExpressionAnalysisContex
             String nodeName = "collection_" + name;
             List<Symbol> outerArguments = ImmutableList.of(collectSetFunction);
             try {
-                return allocateBuiltinOrUdfFunction(schema, nodeName, outerArguments, context);
+                return allocateBuiltinOrUdfFunction(schema, nodeName, outerArguments, windowDefinition, context);
             } catch (UnsupportedOperationException ex) {
                 throw new UnsupportedOperationException(String.format(Locale.ENGLISH,
                     "unknown function %s(DISTINCT %s)", name, arguments.get(0).valueType()), ex);
             }
         } else {
-            return allocateBuiltinOrUdfFunction(schema, name, arguments, context);
+            return allocateBuiltinOrUdfFunction(schema, name, arguments, windowDefinition, context);
+        }
+    }
+
+    private WindowDefinition getWindowDefinition(Window window, ExpressionAnalysisContext context) {
+        List<Symbol> partitionSymbols = new ArrayList<>(window.getPartitions().size());
+        for (Expression partition : window.getPartitions()) {
+            partitionSymbols.add(convert(partition, context));
+        }
+
+        OrderBy orderBy = OrderyByAnalyzer.analyzeSortItems(window.getOrderBy(), sortKey -> convert(sortKey, context));
+
+        WindowFrameDefinition windowFrameDefinition = null;
+        if (window.getWindowFrame().isPresent()) {
+            WindowFrame windowFrame = window.getWindowFrame().get();
+            FrameBound start = windowFrame.getStart();
+            FrameBoundDefinition startBound = convertToAnalyzedFrameBound(context, start);
+
+            FrameBoundDefinition endBound = null;
+            if (windowFrame.getEnd().isPresent()) {
+                endBound = convertToAnalyzedFrameBound(context, windowFrame.getEnd().get());
+            }
+
+            windowFrameDefinition = new WindowFrameDefinition(windowFrame.getType(), startBound, endBound);
+        }
+
+        return new WindowDefinition(partitionSymbols, orderBy, windowFrameDefinition);
+    }
+
+    private FrameBoundDefinition convertToAnalyzedFrameBound(ExpressionAnalysisContext context, FrameBound start) {
```

Comment:
would probably make sense to rename `start` as it is used for both start and end